### ISSUE TITLE
Structure of models directory

### ISSUE THREAD

**Issue by mbaudis** [ga4gh-beacon/beacon-v2-Models#78]
The current v2 models repository contains a root `BEACON-V2-Model` directory, in which one finds directories for the individual schemas and their endpoints. The schemas themselves are named e.g. `biosamples/defaultSchema.json`. Also, the `BEACON-V2-Model` directory contains the overall configuration, map and endpoints files.

To provide a clearer blueprint for extensions and separation between default and alternative schemas it would IMO make sense to re-structure `models` which could easily be done as part of the move to a unified Beacon repository https://github.com/ga4gh-beacon/beacon-v2-Models/issues/76. A possible layout could look like this:

```
beacon
  |
  |-- framework
  |-- models
  |      |-- src
  |      |      |-- beaconConfiguration.yaml
  |      |      |-- beaconMap.yaml
  |      |      |-- endpoints.yaml
  |      |      |-- alternativeModels
  |      |      |      |-- phenopackets
  |      |      |      |      |-- schema.yaml
  |      |      |      |      |-- endpoints.yaml
  |      |      |      |      |-- filteringTerms.yaml
  |      |      |      |      |-- examples
  |      |      |      |-- ...otherAlternativeSchema...
  |      |      |
  |      |      |-- defaultModels
  |      |             |-- analyses
  |      |             |      |-- schema.yaml
  |      |             |      |-- endpoints.yaml
  |      |             |      |-- filteringTerms.yaml
  |      |             |      |-- examples
  |      |             |-- biosamples
  |      |             |      |-- schema.yaml
  |      |             |      |-- endpoints.yaml
  |      |             |      |-- filteringTerms.yaml
  |      |             |      |-- examples
  |      |             |-- ...
  |      |
  |      |-- json
  |      |      |-- beaconConfiguration.json
  |      |      |-- beaconMap.json
...
  |-- docs
  |-- (tools / bin ...)
```

Also note the change from `defaultSchema....` to `schema....`.

Such a restructuring would also allow to have a template which includes the Beacon default model - or rather include the example as part of the standard repo.

Other options would be:
B) as above, but split at the root:
```
  |-- models
  |      |-- alternativeModels
  |      |      |-- src
  |      |      ...
  |      |      
  |      |-- defaultModels
  |      |      |-- src
  |      |      ...
```
C) assume that only the entities in the default model will be allowed, with e.g. `biosampleMySchema.yaml` inside the `biosamples` - not flexible & potentially confusing; also may clash w/ concepts where the general data model is different (e.g. Phenopackets can be representedon the `individuals` level but this isn't correct _senso stricto_).
D) add alternative models to the current structure (e.g. `biosample` and `myBiosample`) - possible but again no separation from Beacon default & possible proliferation of naming issues...

Feedback, please!

**Comment by Tom-Shorter** [comment]
I can't see any reasons why a structure change like you proposed in A or B would cause any problems. I prefer A over B simply due to having files of the same type grouped together which makes it easier to work with that file type in a specialist editor. Option B would mean singling out the correct file format from the directory structure for each entry type, still perfectly fine to do but not as convenient as A.

I would be strongly against option C, we don't know the data held by people who would want to use beacon, or the domain they work in so we must maintain flexibility.

I am less against option D but I still don't think it's the best choice. If someone forks the repo and adds in their own custom entry types/schemas then having a clear distinction between the beacon default schemas and their own custom schemas through the structure alone would be helpful.

Would it make sense to move the beaconConfiguration, map and endpoints files into the framework (merging when already within the framework) as well? These files describe how the beacon instance has been created and what is available to use, this fits more in line with the framework of the beacon than the models IMO.

**Comment by jrambla** [comment]
@mbaudis I see some issues in your assumptions:

1. The `models` repo is plural because initially it contained more than one model, one of them being the Beacon-v2-Model. As some of them became outdated and, at that moment, it seems that every model would have its own repo, I've removed the other models, but didn't dare to rename the repo, neither to remove the root folder.
2. The alternativeSchemas are just that: "schemas", they are not (full)"models"
3. I envision alternativeSchemas being custom for every Beacon instance, some would support Phenopackets, others maybe also OMOP or you-name-it. Therefore, I've always assumed that they are not part of the Beacon v2 Model, and they will no be in that github repo
4. the only exception to that is happening in the Framework, for the `service-info` schema, that is mandatory for all GA4GH services
5. I've found the name `defaultSchema.json` (or yaml) useful to directly refer to the file in the documentation, in conversations and everywhere in contrast with alternative schemas... I understand the `schema.xyz` could be cleaner but also harder to reference. No feelings or personal attachments from my side on this.
6. In terms of additional models, I was not expecting people to add them to the core repository, but creating their own repos and referring to the Framework (or to parts of the Bv2 Model if they simply extend it)

In summary, it is hard for me to comment on the suggested options as I'd visualized in a different way. 
@mbaudis could you comment on your views on the above points so we can further align?

**Comment by mbaudis** [comment]
@jrambla Let me try:

1. plural etc.: Yes, understood; and in the structure above it is slightly off - should be `models -- defaultModel` not plural for the latter (fixed it). But the theoretically improved modularity of having individual models in their own repos is IMO trumped by the confusion and impracticality of having "things Beacon" all over the place.
2. Sure.
3. Yes; the Beacon repo will just represent the default and possibly a blueprint for an alternative model. Complete models can live in their own repos or (more probable) in modified forks of the Beacon repo.
4. Sure.
5. Also no strong opinion there. Only "default" gives the impression that there could be something else in the same space which I guess would clash with the general setup? Not 100% sure; but if no other schema should be provided for an entity in a given model -> this is the default by lack of choice.  I'm a reductionist... `¯\_(ツ)_/¯`
6. See 3. I am not too strongly in favour of `alternativeModel` necessarily here though I see this as a good/preferable option; and the "template model" could be in the repo.

**Comment by jrambla** [comment]
I don't have strong opinions either, just trying to anticipate how all the Beacon v2 ecosystem will unfold and organizing according to the "less resistance path".
I'm expecting to have models coming from different domains than healthcare and genomics, like imaging or proteomics or lists of registries... if I imagine that in a single repo or branched repo, it seems strange to me. On the contrary, I can see an imaging community having a Beacon template for them close to their popular github repos. In such scenario, I believe that cloning the whole unified repo would not make sense for them.
Having said so, one option doesn't preclude the other, and the "other domain" community could probably just point to the relevant folder... although that seems less powerful and elegant to me.
So let's follow the one-step-at-a-time principle and move forward with the unified approach.

**Comment by jrambla** [comment]
I'm not solid in all the details of the above suggestion, hence, let's discuss them as they unfold.

**Comment by mbaudis** [comment]
@jrambla Question: Do we need to have the beaconConfiguration/...map/endpoints sit inside a model (in contrast to my example A)? IMO they shouldn't be part of single models, to allow extension. Also, as @Tom-Shorter said:

> Would it make sense to move the beaconConfiguration, map and endpoints files into the framework (merging when already within the framework) as well? These files describe how the beacon instance has been created and what is available to use, this fits more in line with the framework of the beacon than the models IMO.

OTOH, the fastest solution would be just to use the complete separation right now & fix later? WDYT?

**Comment by Tom-Shorter** [comment]
Following on from the meeting I agree that within the unified repo we don't need an `alternativeModels` directory. This should be the central beacon repo where you find all things that are standard beacon. 

We should encourage people to create alternative models but this will likely be done in 2 methods:

- A user will fork the beacon repo and add in their own models using whatever structure they want and then use that repo as a template for their own beacon instance
- Communities/Networks will create independent repos which contain beacon models specifically for their domain (let the experts in a domain handle their own repos/models instead of trying to deal with them within the central repo)

Both methods should be encouraged, we can also create a section within the readme for the beacon repo which links out to the domain specific repos where appropriate.

**Comment by mbaudis** [comment]
Probably best - in fact, I thought that the alternativeModel would be just in documentation or possibly a the template model.

Still a question for me if the configuration, endpoints... should live within the model or outside (and where?).

**Comment by delocalizer** [comment]
Perhaps not the exact issue being discussed here but still under the broad title of this issue "Structure of models directory" — it's not clear to me (as someone outside the project) what is the difference between the top level directories in this repo

`BEACON-V2-Model`

and 

`BEACON-V2`

?

They both appear to contain JSON schema files with similar but not identical content.

Is this difference explained anywhere in the documentation?
